### ISSUE TITLE
Create Texture Upload and Download

### ISSUE THREAD

**Issue by redorav** [redorav/corsair-engine#13]
# Model

Texture uploading and downloading needs to happen explicitly for DX12-class APIs.

Having an auxiliary command buffer that lives on the render device and does uploads and downloads allows us to defer the upload phase until the frame is about to start, or we can send the command buffer early and get another one, etc It can also deal with using the copy queue. Perhaps a texture uploader class would be useful, in that the code would be platform-independent and would deal with scheduling and other complexities. This upload should handle buffers as well.

There would be two APIs, one scheduled on the render device, another on a command buffer directly. They have different purposes.

- The render device one is meant to be used outside the flow of the frame, and can be more disruptive. It's meant mainly for tooling and non performance-critical bits of code. It can be immediate or deferred (waiting on a lambda, etc)
- The command buffer one is more straightforward, as it follows the GPU timeline directly. It cannot be immediate or it will disrupt the natural flow of the frame. We can e.g. copy a texture to a buffer, add to a queue and pick up the result later in a lambda (such as save the texture to disk)

# Resource States

One of the most complicated issues of explicit APIs is having to track the state of resources. The render graph can track those resources, but outside of that we need to make certain guarantees.

- During the execution and flow of the render graph, every resource state is known
- If the resource is not tracked by the graph, or we are outside the execution of the graph, the resource is assumed to be in a default state
- All operations that change the state of a resource must restore it to the default state after they finish. This needs to be a strong guarantee or we'll get validation errors
- For example, the render graph will always put a tracked resource in its default state at the end

# Requirements

- Need to guarantee texture isn't in use by GPU
- Need an immediate mode (e.g. downloading texture to immediately save it)

## Consoles

- Direct access to memory. No need to copy, it's just there

 ## PC

- Need to download/upload the texture to the CPU via PCIe. Possible latency

## Implementation Problems

- State of texture is not known (for barriers)
- Need to return memory, then dispose of it or keep it around - fixed by returning a buffer
- Need to have option of waiting or not

# Scenarios

## 1. Download pixel memory immediately
Console: Read immediately
PC: Queue download, wait for idle, read immediately

```cpp
// Download texture and read immediately
renderDevice->DownloadTextureImmediate(ICrTexture* texture, [](void* memory, uint64_t size)
{
    // Copy or do behavior here
});

// Creates a CPU-visible buffer
// Puts command in render device command buffer to copy texture into buffer
// Waits for device to finish
// Returns buffer
CrGPUBufferHandle buffer = renderDevice->DownloadTextureImmediate(texture);
```

## 2. Download texture and read whenever it's ready (e.g. continuous read)
Both: Insert a callback that executes when memory is ready

```cpp
renderDevice->DownloadTextureQueue(ICrTexture* texture, [](void* memory, uint64_t size)
{
    // Copy or do behavior here
});

// Creates a buffer
// Puts command in render device command buffer to copy texture into buffer
// Puts a request in the transfer queue
// Loops through the transfer queue every frame, and if finished, executes callback
renderDevice->DownloadTextureQueue(texture, [](ICrHardwareBuffer* buffer)
{
    // Copy or do behavior here
});

// TODO An alternative is to have a BeginDownload() and a wait for download.
```

## 3. Copy pixel memory onto a texture

```cpp
// Transitions texture to dst
// Copies buffer into texture
// Transitions texture into original state
// Optionally submits and waits for device to finish
renderDevice->UploadTextureImmediate(buffer, texture);

Uploading textures to GPU from CPU. Has to have option to stream it directly
On consoles it would just copy it directly
On PC it would queue it at the end of the function so it gets uploaded/swizzled
This API is unable to properly issue barriers
renderDevice->UploadTexture(ICrTexture* texture, [](uint8* stagingBufferMemory)
{
// Serialize texture here into stagingBufferMemory
});

// Begin texture upload
// This creates a temporary buffer on NUMA, maps the texture directly on UMA
// Returns a pointer, so that we can populate the buffer as fast as we see fit
uint8_t* textureData = renderDevice->BeginTextureUpload(texture);

// Populate textureData
for(int w = 0; w < width; ++w) ...

renderDevice->EndTextureUpload(texture);
```
